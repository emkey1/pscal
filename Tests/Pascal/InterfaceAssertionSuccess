program InterfaceAssertionSuccess;

type
  IGreeter = interface
    procedure Greet(const msg: string);
  end;

  TConsoleGreeter = record
    procedure Greet(const msg: string); virtual;
  end;

procedure TConsoleGreeter.Greet(const msg: string);
begin
  writeln('greet: ', msg);
end;

var
  instance: ^TConsoleGreeter;
  iface: IGreeter;
  recovered: ^TConsoleGreeter;
  recoveredViaIs: ^TConsoleGreeter;
begin
  new(instance);
  iface := IGreeter(instance);
  recovered := iface as TConsoleGreeter;
  if recovered = instance then
    writeln('as ok')
  else
    writeln('as mismatch');
  recoveredViaIs := iface is TConsoleGreeter;
  if recoveredViaIs = instance then
    writeln('is ok')
  else
    writeln('is mismatch');
  iface.Greet('through interface');
  dispose(instance);
end.
