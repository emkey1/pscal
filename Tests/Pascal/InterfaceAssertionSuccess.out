as ok
is ok
greet: through interface
